(TK-212) Error when service version already exists

MSLilah opened the pull request:
Throw an exception when a user tries to register multiple
callbacks for a service at the same service level.

MSLilah committed:
(TK-212) Error when service version already exists

Throw an exception when a user tries to register multiple
callbacks for a service at the same service level.

cprice404 commented on test/puppetlabs/trapperkeeper/services/status/status_core_test.clj:
```diff
@@ -15,4 +15,17 @@
+                  "version causes an error")
+      (is (thrown? IllegalStateException
+                   (update-status-context status-fns "foo"
+                                          "1.2.0" 2 (fn [] "foo repeat")))))))
```
This raises an interesting question: does it even make sense to allow registration of more than one `service-version` for a service?  I kind of think the answer is "no".  So maybe we should check for that and error out as well?

rlinehan replied:
I would think that we do want to allow registration for more than one `service-version`? For example, we make some changes to puppet server's status function in puppet server 2.5.0 and thus have a puppet server status function v2, but we still want to support puppet server status function v1. Thus we would have in puppet server's init

``` clojure
(register-callback "puppet-server" "2.5.0" 1 puppetserver-status-v1)
(register-callback "puppet-server" "2.5.0" 2 puppetserver-status-v2)
```

cprice404 replied:
We definitely want to support more than one `status-version`, totally agree.  But `service-version`?  That'd be more like if we registered callbacks for `puppet-server` version `2.4.9` and `2.5.0` in the same process... which seems... craycray?  I might be missing something but so far I haven't been able to imagine a situation where that seems like it would make sense.

rlinehan replied:
Ah, I see. I was completely misreading what you were suggesting (I was thinking you were saying not more than one status function registered per service-version, but really what you were saying is no matter how many status functions are registered by a service, they must all have the same service-version). That does seem like it would be really bizarre, and I can see throwing an error for that.

MSLilah replied:
That should be a really easy check, I'll go ahead and add it.

rlinehan commented:
:+1: this looks good to me!

MSLilah committed:
(TK-212) Throw on differing service version

Throw an IllegalStateException if a user tries to register
a callback for a service which already has one or more callbacks
registered with a differing service version.

MSLilah committed:
(MAINT) Add validate-callback-registration fn

Add a new function, validate-callback-registration, which
checks if a status function can be registered with the given
service version and status version, and if not, throws an
exception.

MSLilah commented:
@rlinehan @cprice404 I've added some error checking for the case in which a user tries to register a callback with a different service version than one that already exists for the same service.
